File: src/models/session.rs
use super::complexity::ComplexityLevel;
use super::pattern::Pattern;
use super::playback_state::PlaybackState;
use super::time_signature::TimeSignature;
use std::collections::VecDeque;
use std::time::SystemTime;
use uuid::Uuid;

/// Represents a single user interaction with the tool
#[derive(Debug)]
pub struct PracticeSession {
    /// Unique session identifier
    pub session_id: Uuid,
    /// Currently playing/displayed pattern
    pub current_pattern: Option<Pattern>,
    /// Last N patterns generated (max 20 for uniqueness checking)
    pub pattern_history: VecDeque<Pattern>,
    /// Playback tempo in beats per minute (40-300)
    pub tempo_bpm: u16,
    /// Pattern complexity setting
    pub complexity_level: ComplexityLevel,
    /// Time signature for pattern generation
    pub time_signature: TimeSignature,
    /// Current playback status
    pub playback_state: PlaybackState,
    /// Whether current pattern has been shown
    pub pattern_revealed: bool,
    /// Total patterns created this session
    pub patterns_generated: u32,
    /// When session began
    pub session_start: SystemTime,
    /// Most recent user interaction
    pub last_activity: SystemTime,
}

impl PracticeSession {
    /// Create a new practice session
    pub fn new(
        tempo_bpm: u16,
        complexity_level: ComplexityLevel,
        time_signature: TimeSignature,
    ) -> Self {
        Self {
            session_id: Uuid::new_v4(),
            current_pattern: None,
            pattern_history: VecDeque::with_capacity(20),
            tempo_bpm,
            complexity_level,
            time_signature,
            playback_state: PlaybackState::Stopped,
            pattern_revealed: false,
            patterns_generated: 0,
            session_start: SystemTime::now(),
            last_activity: SystemTime::now(),
        }
    }

    /// Add a pattern to history, evicting oldest if at capacity
    pub fn add_to_history(&mut self, pattern: Pattern) {
        if self.pattern_history.len() >= 20 {
            self.pattern_history.pop_front();
        }
        self.pattern_history.push_back(pattern);
    }

    /// Update last activity timestamp
    pub fn update_activity(&mut self) {
        self.last_activity = SystemTime::now();
    }
}

impl Default for PracticeSession {
    fn default() -> Self {
        Self::new(120, ComplexityLevel::Medium, TimeSignature::four_four())
    }
}

File: web/src/models/Session.ts
/**
 * Practice Session Model
 * Represents a user's practice session with state and history
 */

import { Pattern } from './Pattern';
import {
  PracticeSession as IPracticeSession,
  TimeSignature,
  ComplexityLevel,
  PATTERN_CONSTRAINTS,
  TEMPO_CONSTRAINTS,
  TIME_SIGNATURES
} from './types';

export class PracticeSession implements IPracticeSession {
  public readonly sessionId: string;
  public currentPattern: Pattern | null;
  public patternHistory: Pattern[];
  public tempoBpm: number;
  public complexityLevel: ComplexityLevel;
  public timeSignature: TimeSignature;
  public patternRevealed: boolean;
  public patternsGenerated: number;
  public readonly sessionStart: string;
  public lastActivity: string;

  constructor(
    tempoBpm: number = TEMPO_CONSTRAINTS.DEFAULT,
    complexityLevel: ComplexityLevel = 'medium',
    timeSignature: TimeSignature = TIME_SIGNATURES.fourFour
  ) {
    this.sessionId = crypto.randomUUID();
    this.currentPattern = null;
    this.patternHistory = [];
    this.tempoBpm = tempoBpm;
    this.complexityLevel = complexityLevel;
    this.timeSignature = { ...timeSignature };
    this.patternRevealed = false;
    this.patternsGenerated = 0;
    this.sessionStart = new Date().toISOString();
    this.lastActivity = this.sessionStart;
  }

  /**
   * Add a pattern to history, evicting oldest if at capacity
   */
  addToHistory(pattern: Pattern): void {
    if (this.patternHistory.length >= PATTERN_CONSTRAINTS.MAX_HISTORY_SIZE) {
      this.patternHistory.shift(); // Remove oldest
    }
    this.patternHistory.push(pattern);
    this.patternsGenerated++;
    this.updateActivity();
  }

  /**
   * Set current pattern
   */
  setCurrentPattern(pattern: Pattern): void {
    this.currentPattern = pattern;
    this.patternRevealed = false;
    this.addToHistory(pattern);
  }

  /**
   * Update last activity timestamp
   */
  updateActivity(): void {
    this.lastActivity = new Date().toISOString();
  }

  /**
   * Convert session to JSON for storage
   */
  toJSON(): IPracticeSession {
    return {
      sessionId: this.sessionId,
      currentPattern: this.currentPattern ? this.currentPattern.toJSON() : null,
      patternHistory: this.patternHistory.map(p => p.toJSON()),
      tempoBpm: this.tempoBpm,
      complexityLevel: this.complexityLevel,
      timeSignature: this.timeSignature,
      patternRevealed: this.patternRevealed,
      patternsGenerated: this.patternsGenerated,
      sessionStart: this.sessionStart,
      lastActivity: this.lastActivity
    };
  }

  /**
   * Create session from JSON
   */
  static fromJSON(json: IPracticeSession): PracticeSession {
    const session = new PracticeSession(
      json.tempoBpm,
      json.complexityLevel,
      json.timeSignature
    );

    // Override generated values with stored values
    (session as any).sessionId = json.sessionId;
    (session as any).sessionStart = json.sessionStart;
    session.currentPattern = json.currentPattern ? Pattern.fromJSON(json.currentPattern) : null;
    session.patternHistory = json.patternHistory.map(p => Pattern.fromJSON(p));
    session.patternRevealed = json.patternRevealed;
    session.patternsGenerated = json.patternsGenerated;
    session.lastActivity = json.lastActivity;

    return session;
  }
}
